packages/manager/src/features/IAM/hooks/usePermissions.ts

```diff
@@ -50,3 +70,74 @@ export const usePermissions = (
 
   return { permissions: permissionMap } as const;
 };
+
+export type EntityBase = Pick<AccountEntity, 'id' | 'label'>;
+
+export const useEntitiesPermissions = <T extends EntityBase>(
+  entities: T[] | undefined,
+  entityType: EntityType,
+  profile?: Profile,
+  enabled = true
+) => {
+  const queries = useQueries({
+    queries: (entities || []).map((entity: T) => ({
+      queryKey: [entityType, entity.id],
+      queryFn: () =>
+        getUserEntityPermissions(
+          profile?.username || '',
+          entityType,
+          entity.id
+        ),
+      enabled,
+    })),
+  });
+
+  const isLoading = queries.some((query) => query.isLoading);
+  const isError = queries.some((query) => query.isError);
+  const data = queries.map((query) => query.data);
+
+  return { data, isLoading, isError };
+};
+
+export const useQueryWithPermissions = <T extends EntityBase>(
+  query: (
+    params?: Params,
+    filter?: Filter,
+    enabled?: boolean
+  ) => UseQueryResult<T[], APIError[]>,
+  entityType: EntityType,
+  permissionsToCheck: PermissionType[]
+): { data: T[]; hasFiltered: boolean } => {
+  const { data: allEntities } = query();
+  const { data: profile } = useProfile();
+  const { isIAMEnabled } = useIsIAMEnabled();
+  const { data: entityPermissions } = useEntitiesPermissions<T>(
+    allEntities,
+    entityType,
+    profile,
+    isIAMEnabled
+  );
+  const { data: grants } = useGrants(!isIAMEnabled);
+
+  const entityPermissionsMap = isIAMEnabled
+    ? toEntityPermissionMap(
+        allEntities,
+        entityPermissions,
+        permissionsToCheck,
+        profile?.restricted
+      )
+    : entityPermissionMapFrom(grants, entityType as GrantType, profile);
```

Comment:
We won't need to cast entityType as GrantType if we update the `entityPermissionMapFrom` typing to `EntityType`
```suggestion
    : entityPermissionMapFrom(grants, entityType, profile);
```